pkg/storage/wal/segment_test.go

```diff
@@ -186,3 +191,172 @@ func TestMultiTenantWrite(t *testing.T) {
+	outputSizeMB := float64(dst.Len()) / (1024 * 1024)
+	compressionRatio := (1 - (outputSizeMB / inputSizeMB)) * 100
+
+	t.Logf("Input Size: %s\n", humanize.Bytes(uint64(w.InputSize())))
+	t.Logf("Output Size: %s\n", humanize.Bytes(uint64(dst.Len())))
+	t.Logf("Compression Ratio: %.2f%%\n", compressionRatio)
+	t.Logf("Write time: %s\n", compressionTime)
+	sizes, err := r.Sizes()
+	require.NoError(t, err)
+	t.Logf("Total chunks %d\n", len(sizes.Series))
+	t.Logf("Index size  %s\n", humanize.Bytes(uint64(sizes.Index)))
+	sizesString := ""
+	for _, size := range sizes.Series {
+		sizesString += humanize.Bytes(uint64(size)) + ", "
+	}
+	t.Logf("Series sizes: [%s]\n", sizesString)
+}
+
+func TestReset(t *testing.T) {
+	w, err := NewWalSegmentWriter()
+	require.NoError(t, err)
+	dst := bytes.NewBuffer(nil)
+
+	w.Append("tenant", "foo", labels.FromStrings("container", "foo", "namespace", "dev"), []*push.Entry{
+		{Timestamp: time.Unix(0, 0), Line: "Entry 1"},
+		{Timestamp: time.Unix(1, 0), Line: "Entry 2"},
+		{Timestamp: time.Unix(2, 0), Line: "Entry 3"},
+	})
+
+	n, err := w.WriteTo(dst)
+	require.NoError(t, err)
+	require.True(t, n > 0)
+
+	copyBuffer := bytes.NewBuffer(nil)
+
+	w.Reset()
+	w.Append("tenant", "foo", labels.FromStrings("container", "foo", "namespace", "dev"), []*push.Entry{
+		{Timestamp: time.Unix(0, 0), Line: "Entry 1"},
+		{Timestamp: time.Unix(1, 0), Line: "Entry 2"},
+		{Timestamp: time.Unix(2, 0), Line: "Entry 3"},
+	})
+
+	n, err = w.WriteTo(copyBuffer)
+	require.NoError(t, err)
+	require.True(t, n > 0)
+
+	require.Equal(t, dst.Bytes(), copyBuffer.Bytes())
+}
+
+func BenchmarkWrites(b *testing.B) {
+	files := testdata.Files()
+	lbls := []labels.Labels{}
+	generators := []*testdata.LogGenerator{}
+
+	for _, file := range files {
+		lbls = append(lbls, labels.FromStrings("filename", file, "namespace", "dev"))
+		lbls = append(lbls, labels.FromStrings("filename", file, "namespace", "prod"))
+		g := testdata.NewLogGenerator(b, file)
+		generators = append(generators, g, g)
+	}
+	inputSize := int64(0)
+	data := []struct {
+		tenant  string
+		labels  string
+		lbls    labels.Labels
+		entries []*push.Entry
+	}{}
+	for inputSize < 5<<20 {
+		for i, lbl := range lbls {
+			more, line := generators[i].Next()
+			if !more {
+				continue
+			}
+			inputSize += int64(len(line))
+			data = append(data, struct {
+				tenant  string
+				labels  string
+				lbls    labels.Labels
+				entries []*push.Entry
+			}{
+				tenant: "tenant",
+				labels: lbl.String(),
+				lbls:   lbl,
+				entries: []*push.Entry{
+					{Timestamp: time.Unix(0, int64(i*1e9)), Line: string(line)},
+				},
+			})
+
+		}
+	}
+
+	dst := bytes.NewBuffer(make([]byte, 0, inputSize))
+
+	pool := sync.Pool{
+		New: func() interface{} {
+			writer, err := NewWalSegmentWriter()
+			if err != nil {
+				panic(err)
+			}
+			return writer
+		},
+	}
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		writer := pool.Get().(*SegmentWriter)
+
+		dst.Reset()
```

Comment:
This benchmark isn't running in parallel (since dst is reused). You could remove the pool for writers from the benchmark, or add a pool for dst and run the whole thing in parallel :)